src/extras.jl

```diff
@@ -113,6 +113,10 @@ function cut(x::AbstractArray{T, N}, breaks::AbstractVector;
              extend::Bool=false,
              labels::Union{AbstractVector{<:AbstractString},Function}=default_formatter,
              allow_missing::Bool=false) where {T, N}
+    if !allunique(breaks)
+        throw(ArgumentError("all breaks must be unique"))
```

Comment:
I think it is OK to just throw an error. I would just highlight the user to pass custom labels as a vector or as a function so that the user knows how to fix the problem. Then it might be good to export `default_formatter` (with some better name) so that then the user can reuse it for users own formatting. The reason is that when I asked people how they would resolve it is to prepend the bin number before the range eg something like "1: [1,2]"` or similar, so it would be nice to be able to reuse a default formatter. Also then probably it would be better to pass data into a formatter as a named tuple as it is easier to handle using anonymous functions, because you could write something like:
```
x - > string(x.i) * ": " * default_formatter(x)
```
(those are just loose thoughts - the key point is to prompt user how to solve the problem not only that there is a problem)

The key issue is that often reducing a number of bins is not an option as you want to eg. have high granularity on the tail of the distribution even if it is concentrated in the center of the mass.

Reply:
Mmm, your reply seems to indicate that we should support this with a simpler solution. Exporting the default formatter would be useful anyway, but that's relatively complex to use. And that doesn't avoid the error about duplicate breaks.

So maybe we should add an argument to allow duplicate breaks as long as they have distinct labels. When a number of quantiles is specified rather than explicit breaks, maybe the formatter should print the quantile number by default, and allow duplicate breaks?

Reply:
> When a number of quantiles is specified rather than explicit breaks, maybe the formatter should print the quantile number by default, and allow duplicate breaks?

This was my initial thought, but I did not want to add work for you with this (as this can be achieved with a custom formatter). The only problem with this is that the string `cut` produces are usually used in plots as x-axis, when quantile number is not needed. So maybe we can add a kwarg if a quantile number should be added or not by default_formatter as a half measure (that would be `false` by default and the error message could indicate that if `default_formatter` is used then it can be set to `true` to avoid the problem)

Reply:
Actually I was thinking the quantile number is an interesting information when you requested quantiles (rather than specific break points). Sometimes you don't even want to see the actual values. What do you think?

Reply:
I think that it is useful, but it would be good to be able to opt-out. As noted - if you do a histogram you typically want x-axis just to give the ranges on the plot.

Reply:
Yes, of course. Maybe we can just provide two formatters, keeping the current one when you pass a vector of breaks, but using another one when you pass a number of quantiles?

Reply:
Before we move forward with this discussion - it seems that actually it seems we have even a more different situation:
```
julia> Random.seed!(1234); z = shuffle!([zeros(10); ones(10)])
20-element Array{Float64,1}:
 0.0
 0.0
 1.0
 1.0
 0.0
 0.0
 0.0
 1.0
 0.0
 1.0
 1.0
 0.0
 1.0
 1.0
 0.0
 0.0
 1.0
 0.0
 1.0
 1.0

julia> zc = cut(z, 10)
20-element CategoricalArray{String,1,UInt32}:
 "[0.0, 0.5)"
 "[0.0, 0.5)"
 "[1.0, 1.0]"
 "[1.0, 1.0]"
 "[0.0, 0.5)"
 "[0.0, 0.5)"
 "[0.0, 0.5)"
 "[1.0, 1.0]"
 "[0.0, 0.5)"
 "[1.0, 1.0]"
 "[1.0, 1.0]"
 "[0.0, 0.5)"
 "[1.0, 1.0]"
 "[1.0, 1.0]"
 "[0.0, 0.5)"
 "[0.0, 0.5)"
 "[1.0, 1.0]"
 "[0.0, 0.5)"
 "[1.0, 1.0]"
 "[1.0, 1.0]"

julia> zc.refs
20-element Array{UInt32,1}:
 0x00000004
 0x00000004
 0x00000008
 0x00000008
 0x00000004
 0x00000004
 0x00000004
 0x00000008
 0x00000004
 0x00000008
 0x00000008
 0x00000004
 0x00000008
 0x00000008
 0x00000004
 0x00000004
 0x00000008
 0x00000004
 0x00000008
 0x00000008

julia> levels(zc)
8-element Array{String,1}:
 "[0.0, 0.0)"
 "[0.0, 0.0)"
 "[0.0, 0.0)"
 "[0.0, 0.5)"
 "[0.5, 1.0)"
 "[1.0, 1.0)"
 "[1.0, 1.0)"
 "[1.0, 1.0]"

julia> droplevels!(zc); levels(zc)
2-element Array{String,1}:
 "[0.0, 0.5)"
 "[1.0, 1.0]"
```

So - maybe it is enough to ensure we `droplevels!` after cutting? You know the sources better so maybe you can quickly validate this assumption? But it seems to me that we actually do not use those "extra duplicate" levels the code produces.